Web dev proxy

salemiranloye opened the pull request:
Adding a development Proxy and dedicated Web configuration File

<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Flutter's current web development configuration relies on CLI arguments and does not have a development proxy. This PR adds a development proxy to flutter and a designated web_dev_config.yaml where web configuration settings are loaded from.

Issues:
https://github.com/flutter/flutter/issues/170834

Document:
[HERE](https://docs.google.com/document/d/1Ud9D3F0GxB5Ocoo5NnAy7PH5oo3qxvALxUlAXMCANJ0/edit?usp=sharing)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

Co-authored-by: Sydney Bao <s.bao2115@gmail.com>
Co-authored by: Salem Iranloye <salemiranloye@gmail.com>

salemiranloye committed:
Development Proxy & Web Configuration File

Co-authored-by: Sydney Bao <sydneybao@google.com>

Update devfs_config.dart

Sydneybao1 (#30)

* mono commit

Co-authored-by: Sydney Bao <sydneybao@google.com>

* Update devfs_config.dart

* fix devfs_config

---------

Co-authored-by: salemiranloye <salemiranloye@gmail.com>
Co-authored-by: Sydney Bao <sydneybao@google.com>

adding missing licenses

remove variable from comment

remove old test

fix shard test

change hash

formatting

null check

fix null check

change io to file system

Fix test

Fix comma test

move file from isolated

Rename Class

fix analyze

remove source span

kevmoo commented on packages/flutter/test/material/radio_list_tile_test.dart:
```diff
@@ -2285,7 +2285,9 @@ void main() {
     await tester.pumpWidget(
       const MaterialApp(
-        home: Material(child: Center(child: RadioListTile<bool>(value: true, radioSide: side))),
+        home: Material(
```
This change seems unrelated. Maybe a merge issue?

salemiranloye replied:
No content is changed just formatting

kevmoo replied:
I just removed 'em. Unrelated.

kevmoo commented on packages/flutter/test/widgets/tap_region_test.dart:
```diff
@@ -1201,18 +1201,17 @@ void main() {
-                                child: const Text('Test Button'),
-                              ),
-                            ),
+                      builder: (BuildContext context) => Scaffold(
```
This change seems unrelated. Maybe a merge issue?

kevmoo replied:
I just removed 'em. Unrelated.

salemiranloye commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -1098,6 +1086,8 @@ class DebuggingOptions {
     required this.ipv6,
     required this.google3WorkspaceRoot,
     required this.printDtd,
+    // ignore: unused_element_parameter
```
Without it I get the error:

A value for optional parameter 'devConfig' isn't ever given.
Try removing the unused parameter.

devConfig is never set in this file but set in another

salemiranloye replied:
I tried this then it doesn't recognize devConfig as a parameter

ditman replied:
The `ignore` was removed; it was probably needed at some point during development, but not anymore.

kevmoo committed:
undo unrelated formatting

Drop all of the unused packages

drop build_runner, too

move shelf import to isolated

update import for test

another unrelated revert

kevmoo commented on packages/flutter/test/material/radio_list_tile_test.dart:
```diff
@@ -2285,9 +2285,7 @@ void main() {
-        home: Material(
-          child: Center(child: RadioListTile<bool>(value: true, radioSide: side)),
-        ),
+        home: Material(child: Center(child: RadioListTile<bool>(value: true, radioSide: side))),
```
Revert this file change! So weird these keep getting in...

salemiranloye replied:
Okay

loic-sharma commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -1145,6 +1131,7 @@ class DebuggingOptions {
   final bool ipv6;
   final String? google3WorkspaceRoot;
   final bool printDtd;
+  final WebDevServerConfig? devConfig;
```
`devConfig` is a bit of a generic name. I'd prefix this with `web` everywhere to clarify this is for web:

```suggestion
  final WebDevServerConfig? webDevConfig;
```

salemiranloye committed:
Add Old params back to toJson

remove !

formatting

add default to webHeaders

formatting

move !

Rename devconfig to webdevserverconfig

Formatting

rename devconfig & add web bool

refactor IIFE as a method

 host and port non nullable & rename variables

refactor loaddevconfig & add a copy method

removing debugging & adding type to objects

error nit picks

formatting

change web to webMode for consistency

change fromJson method

Development Proxy & Web Configuration File

Co-authored-by: Sydney Bao <sydneybao@google.com>

Update devfs_config.dart

Sydneybao1 (#30)

* mono commit

Co-authored-by: Sydney Bao <sydneybao@google.com>

* Update devfs_config.dart

* fix devfs_config

---------

Co-authored-by: salemiranloye <salemiranloye@gmail.com>
Co-authored-by: Sydney Bao <sydneybao@google.com>

adding missing licenses

remove variable from comment

remove old test

fix shard test

null check

change io to file system

Fix test

Fix comma test

move file from isolated

Rename Class

fix analyze

rename field

update tests

rebase

ditman commented:
I'm taking a look at this, but also cc'd @bkonyi and @mdebbar as head honchos of tools and web, respectively!

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+  return fileValue;
+}
+
+Future<WebDevServerConfig> loadDevConfig({
```
This function could live as a `static` method inside the `WebDevServerConfig` class, so it's not a lone global function.

Also if it's a static inside the class, we can call it `loadFromFile` or similar, instead of `loadWebDevServerConfig` which is starting to become a little bit unwieldy! :)

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+  List<String>? browserFlags,
+}) async {
+  final fs.File devConfigFile = globals.fs.file(devConfigFilePath);
+  var fileConfig = const WebDevServerConfig();
```
Again this pattern of creating a dummy defaults file to get the defaults?

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+      }
+
+      if (!contents.containsKey('server') || contents['server'] is! YamlMap) {
+        // final SourceSpan span = (contents.containsKey('server') && contents['server'] is YamlNode)
+        //     ? (contents['server'] as YamlNode).span
+        //     : contents.span;
```
Remove if we're not adding this feature back.

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+  );
+}
+
+Future<int> resolvePort(int? port) async {
```
The way this method works is a little bit weird, it either tries to find a free port or just checks that the incoming port is in range (but it doesn't check if the port is actually available). Do we even need the port < 0 > 65k validation? Do we even need this method? Is this just a fallback to `findFreePort`?

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+    for (final MapEntry<String, String> entry in headers.entries) {
+      if (fileConfig.headers.containsKey(entry.key)) {
+        globals.printStatus(
+          'Overriding headers "${entry.key}" from $devConfigFilePath ("${fileConfig.headers[entry.key]}") with command-line argument("${entry.value}").',
```
All this logging needs a prefix, like [WebDevServer] or similar.

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+      globals.printTrace(fileConfig.toString());
+    } on YamlException catch (e) {
+      globals.printError('Error: Failed to parse $devConfigFilePath: ${e.message} ${e.span}');
+      rethrow;
```
Are we handling this `rethrow` elsewhere? or should this just do `throwToolExit`, like in the `resolvePort` method?

salemiranloye replied:
Good catch, making it throw the tool exit

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+    fileConfig.port,
+    port != null ? int.tryParse(port) : null,
+  );
+  final String? finalCertPath = _getOverriddenValue<String>(
+    'TLS cert path',
+    fileConfig.https?.certPath,
+    tlsCertPath,
+  );
+  final String? finalCertKeyPath = _getOverriddenValue<String>(
+    'TLS cert key path',
+    fileConfig.https?.certKeyPath,
+    tlsCertKeyPath,
+  );
+  HttpsConfig? finalHttpsConfig;
+  if (finalCertPath != null || finalCertKeyPath != null || fileConfig.https != null) {
+    finalHttpsConfig = HttpsConfig(certPath: finalCertPath, certKeyPath: finalCertKeyPath);
+  } else {
+    finalHttpsConfig = null;
+  }
```
This looks like a private `HttpsConfig? _loadHttpsConfig` method, so we can split `loadWebDevServerConfig` into smaller chunks.

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+  }
+}
+
+T? _getOverriddenValue<T>(String filedName, T? fileValue, T? cliValue) {
```
```suggestion
T? _getOverriddenValue<T>(String filedName, T? fileValue, {T? cliValue}) {
```

If you make at least `cliValue` named (and maybe call it "overrideWith") the invocations of this function would be easier to read (for me)

salemiranloye replied:
Removed the getOrridenvalue was clunky and unnecessary

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+}
+
+Future<WebDevServerConfig> loadDevConfig({
+  String? hostname,
```
I think all these parameters should have the "override" prefix in front of them, so it's clear when invoking this method that we're using them as overrides for whatever is in the `devConfigFilePath`.

ditman commented on packages/flutter_tools/lib/src/web/devfs_proxy.dart:
```diff
@@ -0,0 +1,107 @@
+    final Logger effectiveLogger = logger ?? globals.logger;
+
+    RegExp? proxyPattern;
+    if (source != null && source.isNotEmpty) {
```
A nicer pattern for this method would be to delegate the loading and validation to the subclasses, so:

```dart
interface class ProxyRule {
  static ProxyRule? fromYaml(YamlMap yaml, {Logger? logger}) {
    if (source?.isNotEmpty ?? false) {
      return SourceProxyRule.fromYaml(yaml, logger: logger);
    } else if (regex?.isNotEmpty ?? false) {
      return RegexProxyRule.fromYaml(yaml, logger: logger);
    }
  }
}
```

That way the actual validation for the loading of each proy rule subclass lives within its subclass `static fromYaml` method, and this one doesn't blow up to 50 LoC.


If we want to get fancy, we can add a `static canHandleYaml` method that looks at the yaml and decides if that particular class can handle the method, so all this method in the interface needs to do is to know about all possible subclasses, and delegate the construction to the proper one.

```dart
interface class ProxyRule {
  static ProxyRule? fromYaml(YamlMap yaml...) {
    if (SourceProxyRule.canHandle(yaml)) {
      return SourceProxyRule.fromYaml(...)
    } else if (RegexProxyRule.canHandle(yaml)) {
      return RegexProxyRule.fromYaml(...)
    }
  }
}
```

ditman commented on packages/flutter_tools/lib/src/web/devfs_proxy.dart:
```diff
@@ -0,0 +1,107 @@
+  }
+}
+
+class RegexProxyRule extends ProxyRule {
```
Do we need `extends`? Or can this be `implements`?

ditman commented on packages/flutter_tools/lib/src/web/devfs_proxy.dart:
```diff
@@ -0,0 +1,107 @@
+import '../base/logger.dart';
+import '../globals.dart' as globals;
+
+abstract class ProxyRule {
```
All the public attributes/methods on this file need dart docs.

bkonyi replied:
Can this class be `sealed`? Then we'll have support for exhaustiveness checking.

ditman commented on packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:
```diff
@@ -942,29 +949,40 @@ void main() {
+            'run',
+            '--no-pub',
+            '--no-hot',
+            '--no-resident',
```
Why do we need `--no-resident` now? I hadn't seen this mentioned in the code before? Is this leftover from debugging?

salemiranloye replied:
Leftover from debugging

ditman commented on packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:
```diff
@@ -942,29 +949,40 @@ void main() {
-        final device = FakeDevice(isLocalEmulator: true, platformType: PlatformType.android);
+        final device = FakeDevice(
+          isLocalEmulator: true,
+          platformType: PlatformType.web,
+          targetPlatform: TargetPlatform.web_javascript,
```
Why are we making the default device `web`, instead of injecting a `web` device for web tests?

salemiranloye replied:
These are web header tests

ditman commented on packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:
```diff
@@ -1054,32 +1078,37 @@ void main() {
-              '--web-header',
-              'hurray=flutter,flutter=hurray',
-            ]),
-            throwsToolExit(),
```
This test used to throw tool exit, but now it doesn't. Is it because of the change of the default device? I don't think this test should have changed?

salemiranloye replied:
Why should this throw a tool exit?

salemiranloye replied:
The old test was Throwing an error because it was not handling the run command fully it had not implemented a web runner. The expect is the same just using the webDevServerConfig not the DebuggingOptions

ditman commented on packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:
```diff
@@ -1241,7 +1270,7 @@ void main() {
         throwsToolExit(),
       );
 
-      final DebuggingOptions options = await command.createDebuggingOptions(true);
+      final DebuggingOptions options = await command.createDebuggingOptions();
```
On `drive_test.dart`, `command.createDebuggingOptions(false)` was replaced by no parameters (OK), but here, the call is `createDebuggingOptions(true)`. Why?

ditman commented on packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:
```diff
@@ -1482,6 +1511,7 @@ class FakeDevice extends Fake implements Device {
   @override
   String get displayName => name;
 
+  // THIS IS A KEY FIX
```
`// WHAT IS THIS ABOUT?` :P

salemiranloye replied:
remnants of gemini

ditman commented on packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:
```diff
@@ -1693,3 +1723,41 @@ class FakeAnsiTerminal extends Fake implements AnsiTerminal {
+/// A Fake that implements FeatureFlags and enables web.
+class FakeFeatureFlags extends Fake implements FeatureFlags {
+  @override
+  bool get isWebEnabled => true;
```
This feature flag should most likely be removed from everywhere (cc @kevmoo?). Why do we need this fake?

ditman commented on packages/flutter_tools/test/general.shard/drive/web_driver_service_test.dart:
```diff
@@ -277,13 +278,21 @@ void main() {
       final WebDriverService service = setUpDriverService();
       final device = FakeDevice();
       final webHeaders = <String, String>{'test-header': 'test-value'};
+      final webDevServerConfig = WebDevServerConfig(headers: webHeaders);
```
I think the changes in this test file are much more chill than the ones in `run_test.dart` :P

salemiranloye replied:
Is this good?

ditman replied:
Yep!

ditman commented on packages/flutter_tools/test/general.shard/web/proxy_test.dart:
```diff
@@ -0,0 +1,449 @@
+      'should create SourceProxyRule with source and no replacement',
+      () => testbed.run(() {
+        final yaml =
+            loadYaml('''
```
```suggestion
            loadYaml(r'''
```

? Some strings in the test are raw, others aren't. Please, make them all consistent!

ditman reviewed (commented):
Added a bunch of comments, this is very close!

bkonyi commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+        headers[name] = value;
+      }
+    }
+    if (yaml['host'] is! String && yaml['host'] != null) {
```
We repeat this pattern in multiple places. Consider creating a helper function for this type of check.

bkonyi commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+      }
+    }
+    if (yaml['host'] is! String && yaml['host'] != null) {
+      throwToolExit('Host must be a String. Found ${yaml['host'].runtimeType}');
```
The key is `host` not `Host`, so we should be consistent and use the lowercase string in the error message (here and elsewhere in this file).

bkonyi commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+  if (finalCertPath != null || finalCertKeyPath != null || fileConfig.https != null) {
+    finalHttpsConfig = HttpsConfig(certPath: finalCertPath, certKeyPath: finalCertKeyPath);
+  } else {
+    finalHttpsConfig = null;
```
`finalHttpsConfig` is already `null`, so this `else` statement can be removed.

bkonyi commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+  } else {
+    finalHttpsConfig = null;
+  }
+  final finalHeaders = <String, String>{};
```
Ubernit: consider adding `fileConfig.headers` to the map when we create it:

```dart
final finalHeaders = <String, String>{
  ...fileConfig.headers,
};
```

bkonyi commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+
+import '../base/common.dart';
+import '../base/file_system.dart' as fs;
+import '../globals.dart' as globals;
```
Nit: we're trying to reduce the reliance on `globals` for various reasons. For things like file system and logger instances, it's preferred that objects that need access to those resources take them in as parameters instead, particularly when we're adding new classes that don't already rely on `globals`.

bkonyi commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+  final finalHeaders = <String, String>{};
+  finalHeaders.addAll(fileConfig.headers);
+  if (headers != null && headers.isNotEmpty) {
+    for (final MapEntry<String, String> entry in headers.entries) {
```
Consider using patterns here:

```dart
for (final MapEntry(:key, :value) in headers.entries) {
  if (fileConfig.headers.contains(key)) {
    // ...
  }
}


```

salemiranloye replied:
This code was refacotored

bkonyi commented on packages/flutter_tools/lib/src/web/devfs_proxy.dart:
```diff
@@ -0,0 +1,107 @@
+
+  static ProxyRule? fromYaml(YamlMap yaml, {Logger? logger}) {
+    final target = yaml['target'] as String?;
+    final source = yaml['source'] as String?;
```
What happens if `source` and `regex` are both provided? It looks like the `SourceProxyRule` will always take precedence, but should this be an error?

SydneyBao replied:
It's fixed! Thank you

bkonyi reviewed (commented):
First pass :)

salemiranloye commented on packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:
```diff
@@ -352,7 +336,8 @@ Please provide a valid TCP port (an integer between 0 and 65535, inclusive).
         );
         Uri url = await device!.devFS!.create();
-        if (debuggingOptions.tlsCertKeyPath != null && debuggingOptions.tlsCertPath != null) {
+        if (debuggingOptions.webDevServerConfig?.https?.certKeyPath != null &&
```
Nice catch, also what is the copy factory?

ditman replied:
> what is the copy factory?

Sorry, I should have worded that better. by "copy factory" I meant a factory method that copies from another instance of the WebDevServerConfig class:

See:

* https://wiki.c2.com/?FactoryMethod
* https://wiki.c2.com/?CopyConstructor

Specifically in Dart, I'd do this as a `static Type? copyFrom(Type other)` (because I think we needed to return `null`), or a [Factory constructor](https://dart.dev/language/constructors#factory-constructors).

salemiranloye commented on packages/flutter_tools/lib/src/isolated/web_asset_server.dart:
```diff
@@ -216,28 +211,35 @@ class WebAssetServer implements AssetReader {
     required Platform platform,
     bool shouldEnableMiddleware = true,
   }) async {
+    final String effectiveHost = webDevServerConfig.host ?? 'localhost';
```
Yea, i did this because host was nullable but looking at the design it shouldn't because it has default values

salemiranloye commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+        throwToolExit('Headers must be a List of maps. Found ${yaml['headers'].runtimeType}');
+      }
+      final headersList = yaml['headers'] as YamlList;
+      for (final dynamic item in headersList) {
```
Nice Catch

salemiranloye commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,261 @@
+
+  factory WebDevServerConfig.fromYaml(YamlMap yaml) {
+    final headers = <String, String>{};
+    if (yaml['headers'] != null) {
```
How would we define these contraints?

SydneyBao committed:
revert tap_region

update pubspec

Formatting

fix test

fix vmFactory Test

fix test & change default

salemiranloye committed:
Fix Files Reverted in Squash

SydneyBao committed:
update proxy

ditman reviewed (commented):
This is looking good to me mostly, there's some unaddressed questions that I'd like an answer to (I left mentions for @SydneyBao / @salemiranloye in most places).

Also the changes to `packages/flutter_tools/pubspec.yaml` look not needed? For example, where are we using `mockito` in this PR?

SydneyBao committed:
update pubspec

salemiranloye committed:
Allow host and port to be nullable in file but not object nullable

salemiranloye committed:
update packages

salemiranloye committed:
undo changes to root pubspec

ditman commented on packages/flutter_tools/lib/src/commands/drive.dart:
```diff
@@ -304,7 +305,17 @@ class DriveCommand extends RunCommandBase {
+            overrideTlsCertKeyPath: stringArg('web-tls-cert-key-path'),
+          )
+        : null;
+    final web = webDevServerConfig != null;
```
Thanks! Great improvements there, thanks!! 🙏

salemiranloye committed:
downgrade dwds package

salemiranloye committed:
Remove unneed packages

salemiranloye committed:
update hash

salemiranloye committed:
Undo uneeded change

bkonyi reviewed (commented):
Just a couple more comments in addition to the unresolved comments from my last review, mostly WRT defining and using constants for collection keys.

One suggestion not at all related to the code is to try and avoid using `git push --force` unless absolutely necessary. Force pushing detaches comments from previous commits and makes it so the "View changes since last review" button doesn't work. Not a huge deal, but it saves reviewers a lot of time! 😄

ditman commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,224 @@
+        final serverYaml = contents['server'] as YamlMap;
+        fileConfig = WebDevServerConfig.fromYaml(serverYaml);
+        globals.printStatus(
+          '\n [WebDevServer] Loaded configuration from $webDevServerConfigFilePath',
```
Do we need a `\n` here? The logger should be adding those new lines, not us.

salemiranloye replied:
Good catch

bkonyi commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -504,7 +493,24 @@ class RunCommand extends RunCommandBase {
+        extraHeaders: extractWebHeaders(),
+      );
+    } else {
+      return null;
```
Right, I forgot we had that lint enabled for Flutter. SGTM!

salemiranloye commented on packages/flutter_tools/lib/src/web/devfs_proxy.dart:
```diff
@@ -0,0 +1,190 @@
+    final Logger effectiveLogger = logger ?? globals.logger;
+    if (PrefixProxyRule.canHandle(yaml) && RegexProxyRule.canHandle(yaml)) {
+      effectiveLogger.printError(
+        '[PROXY] Both "prefix" and "regex" are defined in the proxy rule YAML. Only one should be used.',
```
On it!

salemiranloye replied:
@SydneyBao I already did it

salemiranloye committed:
Dependancy Injection & kvalues & improved logging

mdebbar commented on packages/flutter_tools/lib/src/isolated/proxy_middleware.dart:
```diff
@@ -0,0 +1,72 @@
+/// [shelf.Request]. If a rule's pattern matches the request's path, the request is
+/// rewritten and forwarded to the target URI defined in the matching [ProxyRule].
+///
+/// If a proxy request results in a 404 Not Found status from the target, or if an
+/// exception occurs during the proxy attempt, the request is allowed to "fall through"
+/// to the next handler in the Shelf stack.
```
This seems a bit counter-intuitive to me. What's the reason for falling through when a 404 is encountered?

SydneyBao replied:
If the redirected path is invalid, an internal route may be the correct destination for the request. For instance, if a proxy rule has a /dart or /assets pattern, internal routes that match these rules don't immediately crash the application.

mdebbar commented on packages/flutter_tools/lib/src/web/devfs_proxy.dart:
```diff
@@ -0,0 +1,190 @@
+    } else if (PrefixProxyRule.canHandle(yaml)) {
+      return PrefixProxyRule.fromYaml(yaml, effectiveLogger);
+    } else if (RegexProxyRule.canHandle(yaml)) {
+      return RegexProxyRule.fromYaml(yaml, effectiveLogger);
```
I suggest we remove the `PrefixProxyRule`. It's a lot of duplicate code that's just a special case of what `RegexProxyRule` can do already.

At the time of parsing `yaml`, if it has a `prefix` key, we could generate a regex from the prefix: `RegExp(r'^' + prefix)`. Everything else remains the same and should work as is.

What do y'all think?

SydneyBao replied:
I agree that there is a lot of redundant code, but the matches and replace methods are slightly different. 

PrefixProxyRule matches paths that start with the rule while RegexProxyRule checks if the rule matches anywhere in the path. PrefixProxyRule replaces the first occurrence of the rule in the path, but RegexProxyRule replaces all occurrences. We did this to provide more flexibility when using regex and using string rules as a simplified, basic mode (I will add dart docs above these methods to clarify!).

I can move some of the redundant logic into ProxyRule? I know we wanted to limit the logic inside the ProxyRule class, so I can also create a new abstract class that implements ProxyRule?

mdebbar replied:
When the `RegexProxyRule` is created with `RegExp(r'^' + prefix)` it will work exactly like a `PrefixProxyRule` without any other changes. Unless I'm missing something?

`^` will make sure we only match prefixes, and it will only match one occurrence, so `replaceAll` will replace one match at most.

ditman replied:
I think we should leave the `PrefixProxyRule` as a separate class, but I agree we can make it an `extends RegexProxyRule`.

I'd rather *not* remove this, because even in the design phase it _seemed_ that we may want to have a `GlobProxyRule` to match `glob` paths, and I'd rather not overcomplicate the sealed base class.

WDYT @mdebbar?

mdebbar replied:
Discussed offline and decided to leave `PrefixProxyRule` and reduce some of the duplicate code.

mdebbar reviewed (commented):
Thanks for this great refactoring and cleanup!

SydneyBao commented on packages/flutter_tools/lib/src/isolated/proxy_middleware.dart:
```diff
@@ -0,0 +1,72 @@
+    return (shelf.Request request) async {
+      final String requestPath = request.requestedUri.path;
+      for (final rule in effectiveProxy) {
+        if (rule.matches(requestPath)) {
```
Sounds good. Thank you.

salemiranloye committed:
Address PR Commends From mdebbar

salemiranloye committed:
fix: other loadfromfile implimentation

ditman commented on packages/flutter_tools/lib/src/web/devfs_proxy.dart:
```diff
@@ -0,0 +1,190 @@
+class PrefixProxyRule implements ProxyRule {
+  /// Creates a [PrefixProxyRule] with the given [pattern] prefix, [target] URI base,
+  /// and optional [replacement] string.
+  PrefixProxyRule({required String pattern, required String target, String? replacement})
```
This `String pattern` should be `String prefix` however :)

SydneyBao replied:
fixed! thank you!

SydneyBao committed:
udpate logging

mdebbar commented on packages/flutter_tools/test/general.shard/web/proxy_test.dart:
```diff
@@ -0,0 +1,398 @@
+      expect(logger.warningText, contains('Invalid regex pattern'));
+      expect(
+        rule.toString(),
+        r'{pattern: \[invalid, target: http://localhost:8080, replace: null}',
```
```suggestion
        r'{regex: \[invalid, target: http://localhost:8080, replace: null}',
```

SydneyBao replied:
good catch! thank you!

mdebbar commented on packages/flutter_tools/test/general.shard/web/proxy_test.dart:
```diff
@@ -0,0 +1,398 @@
+      expect(rule, isNotNull);
+      expect(
+        rule.toString(),
+        r'{pattern: ^/api/(.*), target: http://localhost:8080, replace: /$1}',
```
```suggestion
        r'{regex: ^/api/(.*), target: http://localhost:8080, replace: /$1}',
```

SydneyBao replied:
good catch! thank you!

mdebbar commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -504,7 +493,36 @@ class RunCommand extends RunCommandBase {
+        certKeyPath: stringArg('web-tls-cert-key-path'),
+      );
+
+      final WebDevServerConfig webDevServerConfig = fileConfig.copyWith(
+        host: stringArg('web-hostname'),
+        port: webPort,
+        https: httpsConfig,
+      );
+      return webDevServerConfig;
```
Are we missing `extractWebHeaders()` somewhere here?

SydneyBao committed:
update proxy tests

salemiranloye committed:
allow null values from yaml file

salemiranloye committed:
Fix formatting

salemiranloye committed:
fix: license check

salemiranloye committed:
fix: comment reference

salemiranloye commented on packages/flutter_tools/lib/src/isolated/web_asset_server.dart:
```diff
@@ -216,28 +211,35 @@ class WebAssetServer implements AssetReader {
     }
     InternetAddress address;
-    if (hostname == 'any') {
+    if (effectiveHost == 'any') {
```
addressed

salemiranloye committed:
remove duplicate logging

ditman commented:
The latest pushes introduced some analyzer issues, PTAL @salemiranloye!

salemiranloye committed:
fix: formatting

salemiranloye committed:
remove obvious type

mdebbar commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -504,7 +493,37 @@ class RunCommand extends RunCommandBase {
+        host: stringArg('web-hostname'),
+        port: webPort,
+        https: httpsConfig,
+        headers: extractWebHeaders(),
```
The headers should be combined:
```suggestion
        headers: <String, String>{...fileConfig.headers, ...?extractWebHeaders()},
```

mdebbar commented on packages/flutter_tools/lib/src/commands/drive.dart:
```diff
@@ -303,8 +304,26 @@ class DriveCommand extends RunCommandBase {
+      host: stringArg('web-hostname'),
+      port: webPort,
+      https: httpsConfig,
+      headers: extractWebHeaders(),
```
Same comment here about combining headers.

mdebbar reviewed (approved):
![LGTM](https://user-images.githubusercontent.com/1278212/240958742-5849e437-2e5f-4b46-ae0a-254395368201.png)

ditman reviewed (approved):
I'm happy with how this turned out and am approving, but please don't merge until @bkonyi approves!

bkonyi reviewed (approved):
Just a couple of minor comments that wouldn't be worth blocking this PR on, but would be great to clean up if you've got the time.

LGTM and thanks for all the hard work! 🥳

SydneyBao commented on packages/flutter_tools/lib/src/web/devfs_config.dart:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    final YamlList? proxyList = _validateType<YamlList>(value: yaml[_kProxy], fieldName: _kProxy);
+    final proxyRules = <ProxyRule>[];
```
done!

SydneyBao committed:
fix comments

salemiranloye committed:
Merge pull request #7 from SydneyBao/final_proxy

fix comments

woprandi commented:
Can we report issue about this feature ?

GreatTux commented:
Issue raised against this change here: https://github.com/flutter/flutter/issues/179014
